### ISSUE TITLE
[network] Setting the MTU on a vxlan interface is throwing an invalid argument error

### ISSUE THREAD

**Issue by shibumi** [systemd/systemd#9254]
**systemd version the issue has been seen with**
> 237

**Used distribution**
 > Ubuntu 18.04

**Expected behaviour you didn't see**
The VXLAN Interface should come up with the right MTU.

**Unexpected behaviour you saw**
The VXLAN interface doesn't come up when a Link section exists with a `MTUBytes` argument.
`journalctl -u systemd-networkd.service` shows in this case:
```
Jun 11 09:14:02 ubuntu-template18 systemd-networkd[4355]: vxlan1309: Could not bring up interface: Invalid argument 
```

**Steps to reproduce the problem**

`/etc/systemd/network/ens160.network`
```
[Match]
Name=ens160

[Link]
MTUBytes=9000

[Network]
Address=xxx.xxx.xxx.100/27
Gateway=xxx.xxx.xxx.126
DNS=xxx.xxx.xxx.2
VXLAN=vxlan1309
```

`/etc/systemd/network/vxlan1309.netdev`:
```
[NetDev]
Name=vxlan1309
Kind=vxlan

[VXLAN]
Id=1309
MacLearning=false
DestinationPort=4789
```

`/etc/systemd/network/vxlan1309.network`:
```
[Match]
Name=vxlan1309

[Link]
MTUBytes=8950

[Network]
Bridge=vxbr1309
```

`/etc/systemd/networkd/vxbr1309.netdev`
```
[NetDev]
Name=vxbr1309
Kind=bridge
```

`/etc/systemd/networkd/vxbr1309.network`
```
[Match]
Name=vxbr1309
```

`networkctl status vxlan1309` : 
```
● 7: vxlan1309
       Link File: /lib/systemd/network/99-default.link
    Network File: /etc/systemd/network/vxlan1309.network
            Type: ether
           State: off (configuring)
          Driver: vxlan
      HW Address: 3a:72:15:cf:c5:19
```

When I remove the `Link`-Section from `vxlan1309.network` the vxlan interfaces comes up as expected.
My Problem is that I want to a VXLAN Interface with MTU 8950 and a physical interface with MTU 9000.
Normally when I would need todo this with the `iproute2`-utils I would just set the MTU 9000 on the physical interface `ens160` and then I would create the vxlan interface and the vxlan interface would have the right MTU `8950` automatically.
`# ip link set ens160 mtu 9000`
`# ip link add vxlan1309 type vxlan id 1309 nolearning dstport 4789 dev ens160`

My suspicion is that the VXLAN interface get not up because it tries to set the MTU 8950, but the physical interface has only MTU 1500, because systemd-networkd fails to set the MTU via udev for the physical interface.

So I guess the main problem here is that the MTU settings get not applied when I restart `systemd-networkd`.

My Questions are:
1. Is defining the MTU in a `link`-Section in the network file even the right way?
2. how to set the MTU without restarting the whole machine? When I reboot the MTU for both interfaces get applied as expected. But without reboot and just restart `systemd-networkd` or `systemd-udev-trigger` the MTU will not get applied. (setting it manually via `ip link set..` works)
3. I have also tried setting the MTU in the netdev file of the vxlan interface. Same result.The MTU will not get applied.

**Comment by ssahani** [comment]
I did certain testing . The vxlan driver returns ```-EINVAL```  if underlying interface MTU is lesser then vxlan MTU. https://elixir.bootlin.com/linux/v4.1.28/source/drivers/net/vxlan.c#L2289
```
[sus@Zeus systemd]$ sudo  ip link set dummy-test mtu 1000
[sus@Zeus systemd]$ sudo  ip link set vxlan1309 mtu 9000
RTNETLINK answers: Invalid argument
[sus@Zeus systemd]$ sudo  ip link set dummy-test mtu 9000
[sus@Zeus systemd]$ sudo  ip link set vxlan1309 mtu 8500
[sus@Zeus systemd]$ ip -d link  show vxlan1309
45: vxlan1309: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 8500 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/ether a6:fc:e5:52:26:b5 brd ff:ff:ff:ff:ff:ff promiscuity 0 
    vxlan id 1309 dev dummy-test srcport 0 0 dstport 4789 nolearning ttl inherit ageing 300 noudpcsum noudp6zerocsumtx noudp6zerocsumrx addrgenmode eui64 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 

```
So the physical interface MTU need to set first the the VXLAN  in the .network files. That means you need to rename the files.

```

# cat dummy1.netdev 
[NetDev]
Name=dummy-test
Kind=dummy
[root@Zeus network]# cat vxlan1309.netdev
[NetDev]
Name=vxlan1309
Kind=vxlan

[VXLAN]
Id=1309
MacLearning=false
DestinationPort=4789
[root@Zeus network]# cat dummy1.network
[Match]
Name=dummy-test

[Link]
MTUBytes=9000

[Network]
VXLAN=vxlan1309
[root@Zeus network]# cat vxlan.network
[Match]
Name=vxlan1309

[Link]
MTUBytes=8000
```

Output
```
[sus@Zeus systemd]$ sudo  ip -d link show vxlan1309
49: vxlan1309: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 8000 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/ether a6:fc:e5:52:26:b5 brd ff:ff:ff:ff:ff:ff promiscuity 0 
    vxlan id 1309 dev dummy-test srcport 0 0 dstport 4789 nolearning ttl inherit ageing 300 noudpcsum noudp6zerocsumtx noudp6zerocsumrx addrgenmode eui64 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 
[sus@Zeus systemd]$ sudo  ip -d link show dummy-test
48: dummy-test: <BROADCAST,NOARP,UP,LOWER_UP> mtu 9000 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/ether fe:57:46:54:5a:2a brd ff:ff:ff:ff:ff:ff promiscuity 0 
    dummy addrgenmode eui64 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 
```

**Comment by shibumi** [comment]
@ssahani Thanks for your answer and tracing back this issue. I don't get why I need to rename the files. At the moment I see not much difference between our both setups. Do you have a specific loading scheme for loading network, netdev and link files? 

At the moment I have something like this:
`ens160.network`
```
[Match]
Name=ens160

[Link]
MTUBytes=9000

[Network]
Address=xxx.xxx.xxx.100/27
Gateway=xxx.xxx.xxx.126
DNS=xxx.xxx.xxx.2
VXLAN=vxlan1309
```

`vxlan1309.network`
```
[Match]
Name=vxlan1309

[Link]
MTUBytes=8950
```

Normally I expect that the physical interface will be get configured first and then the vxlan interface will get set up with the physical interface as carrier.

So where exactly is here the difference to your setup? is it the naming of the network files? So If I would do:

`10-ens160.network` and `99-vxlan1309.network` it would work fine?

My second and last question is: "Could you set the MTU without reboot? I need to reboot for it. Restarting `systemd-networkd` or `systemd-udev-trigger` does not really help here.

**Comment by ssahani** [comment]
sorry I forgot to mention one thing did you tried with  ```ens160``` down then restarting networkd again .
```sudo ip link set dev eth0 down```

**Comment by shibumi** [comment]
@ssahani I can confirm that the interface gets the MTU when I set them down
Here is my current config:
`ens160.network`
```
[Match]                                      
Name=ens160                                  
                                             
[Link]                                       
MTUBytes=9000                                
                                             
[Network]                                    
Address=xxx.xxx.221.99/27                    
Gateway=xxx.xxx.221.126                      
DNS=xxx.xxx.2.5                              
DNS=xxx.xxx.2.6                              
DNS=xxx.xxx.33.138                           
Domains=xxx.xxx.xxx
VXLAN=vxlan1309                              
```

`vxlan1309.network`
```
[Match]               
Name=vxlan1309        
                      
[Link]                
MTUBytes=8950         
                      
[Network]             
Bridge=vxbr1309       
```

What still bothers me is that I need to set the interfaces down. This is a clearly a regression in my opinion.

I mean I can set the MTU directly via `ip link set interface mtu 9000` without setting an interface down. Why is not systemd-networkd able to do this? Because udev handles these things?

At the moment this workaround works:
```
ip link set ens160 down; ip link set vxlan1309 down; systemctl restart systemd-networkd;
```

But I would prefer just a service restart... Shall/Can I open a Feature-Request for this?

**Comment by ssahani** [comment]
Now the MTU is getting set if we trying to up the link that is being fed into the ```UP``` Netlink message. We need to separate that . If the link is up networkd does not try to make it up . This is why MTU setting is not working. Will post a patch

**Comment by shibumi** [comment]
@ssahani great to see a patch in this direction! Thanks for debugging this with me

**Comment by ssahani** [comment]
@shibumi I have referenced the patch. Can you confirm ? [root@Zeus network]# cat ens160.network

```
[Match]
Name=enp0s31f6

[Link]
MTUBytes=9000

[Network]
VXLAN=vxlan1309
[root@Zeus network]# cat vxlan1309.netdev 
[NetDev]
Name=vxlan1309
Kind=vxlan

[VXLAN]
Id=1309
MacLearning=false
DestinationPort=4789
[root@Zeus network]# cat vxlan1309.network
[Match]
Name=vxlan1309

[Link]
MTUBytes=8950

```
```

2: enp0s31f6: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 9000 qdisc fq_codel state DOWN mode DEFAULT group default qlen 1000
    link/ether 8c:16:45:6c:83:b9 brd ff:ff:ff:ff:ff:ff

5: vxlan1309: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 8950 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/ether a6:a6:58:90:fa:d1 brd ff:ff:ff:ff:ff:ff
```

**Comment by shibumi** [comment]
@ssahani yep this looks good. Does it work now without setting the interface down first?

**Comment by ssahani** [comment]
well I need your feedback :)

**Comment by shibumi** [comment]
@ssahani ok let me try this patch out

**Comment by shibumi** [comment]
@ssahani ok I had time to check your commit now.
For this I have used a arch linux vagrant box:

`vagrant init archlinux/archlinux`
`vagrant up && vagrant ssh`

Then I have created the following files:
```
==> 99-default.link <==

==> eth0.network <==
[Match]
Name=eth0

[Link]
MTUBytes=9000

[Network]
DHCP=ipv4

==> vxlan1309.netdev <==
[NetDev]
Name=vxlan1309
Kind=vxlan

[VXLAN]
Id=1309
MacLearning=false
DestinationPort=4789

==> vxlan1309.network <==
[Match]
Name=vxlan1309

[Link]
MTUBytes=8950
```

With default systemd the vxlan interface will not get created and the MTU for the physical interface will not get set.


Then I have installed your systemd version with your commit in it and restarted `systemd-networkd`.
After this the physical interface had the right MTU, but the vxlan interface has not been created.
And the journal log for `systemd-networkd` looks really messy:

```
Jun 15 08:30:49 archlinux systemd-networkd[9814]: Directory "/run/systemd/netif" already exists, but is owned by 0:0 (981:981 was requested), refusing.
Jun 15 08:30:49 archlinux systemd-networkd[9814]: Could not create runtime directory: File exists
Jun 15 08:30:49 archlinux systemd-networkd[9814]: Directory "/run/systemd/netif/links" already exists, but is owned by 0:0 (981:981 was requested), refusin>
Jun 15 08:30:49 archlinux systemd-networkd[9814]: Could not create runtime directory 'links': File exists
Jun 15 08:30:49 archlinux systemd-networkd[9814]: Directory "/run/systemd/netif/leases" already exists, but is owned by 0:0 (981:981 was requested), refusi>
Jun 15 08:30:49 archlinux systemd-networkd[9814]: Could not create runtime directory 'leases': File exists
Jun 15 08:30:49 archlinux systemd-networkd[9814]: eth0: Gained IPv6LL
Jun 15 08:30:49 archlinux systemd-networkd[9814]: lo: Failed to save link data to /run/systemd/netif/links/1: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: eth0: Failed to save link data to /run/systemd/netif/links/2: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: Enumeration completed
Jun 15 08:30:49 archlinux systemd-networkd[9814]: lo: Failed to save link data to /run/systemd/netif/links/1: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: eth0: Failed to save link data to /run/systemd/netif/links/2: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: lo: Failed to save link data to /run/systemd/netif/links/1: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: eth0: Failed to save link data to /run/systemd/netif/links/2: Permission denied
Jun 15 08:30:49 archlinux systemd[1]: Started Network Service.
Jun 15 08:30:49 archlinux systemd-networkd[9814]: lo: Failed to save link data to /run/systemd/netif/links/1: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: eth0: Failed to save link data to /run/systemd/netif/links/2: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: lo: Link is not managed by us
Jun 15 08:30:49 archlinux systemd-networkd[9814]: eth0: IPv6 successfully enabled
Jun 15 08:30:49 archlinux systemd-networkd[9814]: lo: Failed to save link data to /run/systemd/netif/links/1: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: eth0: Failed to save link data to /run/systemd/netif/links/2: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: lo: Failed to save link data to /run/systemd/netif/links/1: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: eth0: Failed to save link data to /run/systemd/netif/links/2: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: lo: Failed to save link data to /run/systemd/netif/links/1: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: eth0: Failed to save link data to /run/systemd/netif/links/2: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: lo: Failed to save link data to /run/systemd/netif/links/1: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: eth0: Failed to save link data to /run/systemd/netif/links/2: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: request_name_destroy_callback n_ref=1
Jun 15 08:30:49 archlinux systemd-networkd[9814]: lo: Failed to save link data to /run/systemd/netif/links/1: Permission denied
Jun 15 08:30:49 archlinux systemd-networkd[9814]: eth0: Failed to save link data to /run/systemd/netif/links/2: Permission denied
```

The last two messages are repeating for over 2minutes..

**Comment by ssahani** [comment]
```request_name_destroy_callback n_ref=1
vboxnet0: Failed to save link data to /run/systemd/netif/links/4: Permission denied
enp0s31f6: Failed to save link data to /run/systemd/netif/links/2: Permission denied
lo: Failed to save link data to /run/systemd/netif/links/1: Permission denied
wlan0: Failed to save link data to /run/systemd/netif/links/3: Permission denied
vboxnet0: Failed to save link data to /run/systemd/netif/links/4: Permission denied
enp0s31f6: Failed to save link data to /run/systemd/netif/links/2: Permission denied
lo: Failed to save link data to /run/systemd/netif/links/1: Permission denied
wlan0: Failed to save link data to /run/systemd/netif/links/3: Permission denied
vboxnet0: Failed to save link data to /run/systemd/netif/links/4: Permission denied
enp0s31f6: Failed to save link data to /run/systemd/netif/links/2: Permission denied
lo: Failed to save link data to /run/systemd/netif/links/1: Permission denied
wlan0: Failed to save link data to /run/systemd/netif/links/3: Permission denied
vboxnet0: Failed to save link data to /run/systemd/netif/links/4: Permission denied
enp0s31f6: Failed to save link data to /run/systemd/netif/links/2: Permission denied
lo: Failed to save link data to /run/systemd/netif/links/1: Permission denied
wlan0: Failed to save link data to /run/systemd/netif/links/3: Permission denied
^C[sus@Zeus build]$ ./systemd-networkd
``` 

```


do sudo ```./systemd-networkd```

**Comment by shibumi** [comment]
mhh I am root. I have no idea why this fails. Setting the MTU works:
```
[vagrant@archlinux ~]$ sudo ip link set eth0 mtu 1500
[vagrant@archlinux ~]$ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 08:00:27:f8:72:e0 brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global dynamic eth0
       valid_lft 86397sec preferred_lft 86397sec
    inet6 fe80::a00:27ff:fef8:72e0/64 scope link 
       valid_lft forever preferred_lft forever
(failed reverse-i-search)`sudo systemctl': ^Cdo ip link set eth0 mtu 1500
[vagrant@archlinux ~]$ sudo -i
[root@archlinux ~]# systemctl restart systemd-networkd
[root@archlinux ~]# ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc fq_codel state UP group default qlen 1000
    link/ether 08:00:27:f8:72:e0 brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global dynamic eth0
       valid_lft 86398sec preferred_lft 86398sec
    inet6 fe80::a00:27ff:fef8:72e0/64 scope link 
       valid_lft forever preferred_lft forever
```

Maybe VM Related?

**Comment by shibumi** [comment]
The MTU for eth0 get set but VXLAN will not get created.. maybe because of the netlink permission errors although I am root (are here kernel capabilities involved?) or maybe because systemd-networkd has the wrong sequence and tries to create the vxlan interface with MTU 8950 before setting the MTU for the physical interface

**Comment by shibumi** [comment]
@ssahani ah ok I found the issue why the vxlan interface does not exist. Because I forgot to mention the physical interface:

==> 99-default.link <==

==> eth0.network <==
[Match]
Name=eth0

[Link]
MTUBytes=9000

[Network]
DHCP=ipv4
**VXLAN=vxlan1309**

==> vxlan1309.netdev <==
[NetDev]
Name=vxlan1309
Kind=vxlan

[VXLAN]
Id=1309
MacLearning=false
DestinationPort=4789

==> vxlan1309.network <==
[Match]
Name=vxlan1309

[Link]
MTUBytes=8950



The bold line was missing.. but what is  odd is that I still get this permission errors in my VM and the DHCP option doesn't work when I set `[Link] MTUBytes=9000` in /etc/systemd/network/eth0.network.

Is this another bug? :)

**Comment by ssahani** [comment]
no I can't reproduce it works for me 

```
NDISC: Received Router Advertisement: flags none preference medium lifetime 65535 sec
NDISC: Invoking callback for 'r'.
enp0s9: Updating address: fe80::58bb:60a2:e3e0:4743/64 (valid forever)
Ignoring option DHCP option 120 while parsing.
DHCP CLIENT (0xa1ce5c37): OFFER
DHCP CLIENT (0xa1ce5c37): REQUEST (requesting)
Ignoring option DHCP option 120 while parsing.
DHCP CLIENT (0xa1ce5c37): ACK
DHCP CLIENT (0xa1ce5c37): lease expires in 11h 59min 58s
DHCP CLIENT (0xa1ce5c37): T2 expires in 10h 29min 57s
DHCP CLIENT (0xa1ce5c37): T1 expires in 5h 59min 58s
enp0s9: DHCPv4 address 192.168.225.214/24 via 192.168.225.1
Setting transient 
```

**Comment by shibumi** [comment]
@ssahani ok then it's just my virtualbox setup.. maybe the virtualbox dhcp doesn't like the jumbo frames aka big MTU. This issue is resolved for me. I think you can open a pull request :)

**Comment by yuwata** [comment]
```
vboxnet0: Failed to save link data to /run/systemd/netif/links/4: Permission denied
enp0s31f6: Failed to save link data to /run/systemd/netif/links/2: Permission denied
lo: Failed to save link data to /run/systemd/netif/links/1: Permission denied
wlan0: Failed to save link data to /run/systemd/netif/links/3: Permission denied
```
I think the error is equivalent to #9369, and will be fixed by #9376.
